common/scala/src/main/scala/whisk/core/containerpool/ContainerClient.scala

```diff
@@ -0,0 +1,212 @@
+   * wait longer than the total timeout (within a small slack allowance).
+   *
+   * @param endpoint the path the api call relative to hostname
+   * @param body the JSON value to post (this is usually a JSON objecT)
+   * @param retry whether or not to retry on connection failure
+   * @return Left(Error Message) or Right(Status Code, Response as UTF-8 String)
+   */
+  def post(endpoint: String, body: JsValue, retry: Boolean)(
+    implicit tid: TransactionId): Future[Either[ContainerHttpError, ContainerResponse]] = {
+
+    //create the request
+    val req = Marshal(body).to[MessageEntity].map { b =>
+      //DO NOT reuse the connection (in case of paused containers)
+      //For details on Connection: Close handling, see:
+      // - https://doc.akka.io/docs/akka-http/current/common/http-model.html#http-headers
+      // - http://github.com/akka/akka-http/tree/v10.1.3/akka-http-core/src/test/scala/akka/http/impl/engine/rendering/ResponseRendererSpec.scala#L470-L571
+      HttpRequest(HttpMethods.POST, endpoint, entity = b).withHeaders(Connection("close"))
+    }
+
+    retryingRequest(req, timeout, retry)
+      .flatMap(response =>
+        response.entity.contentLengthOption match {
+          case Some(contentLength) if response.status != StatusCodes.NoContent =>
+            if (contentLength <= maxResponse.toBytes) {
+              Unmarshal(response.entity.withSizeLimit(maxResponse.toBytes)).to[String].map { o =>
+                Right(ContainerResponse(response.status.intValue, o, None))
+              }
+            } else {
+              truncated(response.entity.dataBytes).map { s =>
+                Right(ContainerResponse(response.status.intValue, s, Some(contentLength.B, maxResponse)))
+              }
+            }
+          case _ =>
+            //handle missing Content-Length as NoResponseReceived
+            //also handle 204 as NoResponseReceived, for parity with HttpUtils client
+            Future { Left(NoResponseReceived()) }
+      })
+      .recover {
+        case t: StreamTcpException => Left(Timeout(t))
+        case t: TimeoutException   => Left(Timeout(t))
+        case t: Throwable          => Left(ConnectionError(t))
+      }
+  }
+  private def retryingRequest(req: Future[HttpRequest],
+                              timeout: FiniteDuration,
+                              retry: Boolean): Future[HttpResponse] = {
+    request(req).recoverWith {
+      case t: akka.stream.StreamTcpException if retry =>
+        val newTimeout = timeout - retryInterval
+        if (newTimeout > Duration.Zero) {
+          akka.pattern.after(retryInterval, as.scheduler)(retryingRequest(req, newTimeout, retry))
+        } else {
+          logging.warn(this, s"POST failed with $t - no retry because timeout exceeded.")
+          Future.failed(t)
+        }
+      case t => Future.failed(t)
+    }
+  }
+
+  private def truncated(responseBytes: Source[ByteString, _],
```

Comment:
May be move it to `object PoolingContainerClient` and then have a test for this logic. Per [current test coverage][1] some flows are not covered

[1]: https://codecov.io/gh/apache/incubator-openwhisk/pull/3812/diff#D10-153

Reply:
There is at least one test already that covers truncation: https://github.com/apache/incubator-openwhisk/pull/3812/files#diff-70e8c471d9056bda26d602a05f7ad091R180

Is codecov.io updated on each build? Can you tell why it wouldn't show in coverage? Theses tests `PoolingContainerClientTests` are using `ContainerClient` directly, no mocks etc, so I'm not sure why the coverage would not reflect?